src/util/getNextThoughtsWithContextChain.js

```diff
@@ -0,0 +1,178 @@
+  return childrenFiltered[0]
+}
+
+const getDerivedCursor = contextChain => {
+  // last of second last item in context chain gives us the current context
+  const context = head(contextChain.slice(-2, -1)[0])
+  const contexts = getContextsSortedAndRanked(context.value)
+  const currentContextTop = head(contextChain)[0].value
+  const matchedContextChain = contexts.find(c => c.context.includes(currentContextTop))
+  const matchedContext = matchedContextChain.context.map(i => ({ value: i, rank: matchedContextChain.rank }))
+
+  return [...matchedContext, context, ...head(contextChain).slice(1)]
+}
+
+const contextViewLookup = (value, context, rank, cursor, contextRanked, contextChain, ignoreChildren) => {
+  const { showHiddenThoughts, thoughtIndex } = store.getState()
+
+  if (context.length === 0 || cursor.length === 0) {
+    return null
+  }
+
+  // if the cursor is at a thought with context view open, move it into context view - jump in
+  if (isValidContextView(pathToContext(cursor))) {
+    const currentThought = head(cursor)
+    const firstChild = firstChildOfContext(value, cursor, thoughtIndex)
+    if (!ignoreChildren && firstChild) {
+      return { nextThoughts: [...firstChild, currentThought], contextChain }
+    }
+    // jump out if there are no context children
+    return getNextThoughtsWithContextChain(value, context, rank, cursor, contextOf(contextRanked), contextOf(contextChain))
+  }
+  // if cursor is within a context view
+  else if (isValidContextView(contextRanked)) {
+    const headContext = contextOf(cursor).map(c => c.value)
+    if (!ignoreChildren && isValidContextView(headContext)) {
+      const derivedCursor = getDerivedCursor(contextChain)
+      const firstChild = firstChildOfThought(derivedCursor || RANKED_ROOT, showHiddenThoughts)
+      if (firstChild) {
+        return { nextThoughts: unroot(cursor.concat(firstChild)), contextChain: [] }
+      }
+    }
+
+    const derivedContextRanked = (contextChain.length === 1 && isValidContextView(contextChain[0])) ? contextChain[0] : contextRanked
+    const derivedContext = derivedContextRanked.map(c => c.value)
+    const nextSibling = nextSiblingContext(value, rank, derivedContext, thoughtIndex, showHiddenThoughts)
+    const contextRankedHead = head(derivedContextRanked)
+
+    if (nextSibling) {
+      return { nextThoughts: [...nextSibling, contextRankedHead], contextChain: contextOf(contextChain) }
+    }
+    // jump out
+    return thoughtViewLookUp(contextRankedHead.value, contextOf(derivedContext), contextRankedHead.rank, contextOf(cursor), contextOf(contextChain), true)
+  }
+  const contextWithNoChildren = contextChain.length === 1 && isContextViewActive(contextChain[0]) && getContextsSortedAndRanked(head(pathToContext(contextChain[0]))).length <= 1
+  return thoughtViewLookUp(value, context, rank, cursor, contextChain, contextWithNoChildren)
+}
+
+const thoughtViewLookUp = (value, context, rank, cursor, contextChain, ignoreChildren) => {
```

Comment:
What's the relationship between `contextViewLookUp` and `thoughtViewLookup`? This deserves a thorough explanation (think 1-2 paragraphs). All I can tell is that you have the recursion split across two functions, but I have no idea what the purpose of this implementation is, how it works, and why you chose it.

Reply:
Added description with example